lib/questionlib.php

```diff
@@ -2594,3 +2076,726 @@ function get_next_version(int $questionbankentryid): ?int {
+                            $a->idnumber = s($category->idnumber);
+                        }
+                        if (!empty($category->questioncount)) {
+                            $a->questioncount = $category->questioncount;
+                        }
+                        if (isset($a->idnumber) && isset($a->questioncount)) {
+                            $formattedname = get_string('categorynamewithidnumberandcount', 'question', $a);
+                        } else if (isset($a->idnumber)) {
+                            $formattedname = get_string('categorynamewithidnumber', 'question', $a);
+                        } else if (isset($a->questioncount)) {
+                            $formattedname = get_string('categorynamewithcount', 'question', $a);
+                        } else {
+                            $formattedname = $a->name;
+                        }
+                        $categoriesarray[$contextstring][$cid] = $formattedname;
+                    }
+                }
+            }
+        }
+        if ($popupform) {
+            $popupcats = array();
+            foreach ($categoriesarray as $contextstring => $optgroup) {
+                $group = array();
+                foreach ($optgroup as $key => $value) {
+                    $key = str_replace($CFG->wwwroot, '', $key);
+                    $group[$key] = $value;
+                }
+                $popupcats[] = array($contextstring => $group);
+            }
+            return $popupcats;
+        } else {
+            return $categoriesarray;
+        }
+    }
+
+    /**
+     * @deprecated since Moodle 4.0 MDL-71585
+     * @see qbank_managecategories\helper
+     */
+    function question_add_context_in_key($categories) {
+        debugging('Function question_add_context_in_key() has been deprecated and moved to qbank_managecategories plugin,
+    Please use qbank_managecategories\helper::question_add_context_in_key() instead.', DEBUG_DEVELOPER);
+        $newcatarray = array();
+        foreach ($categories as $id => $category) {
+            $category->parent = "$category->parent,$category->contextid";
+            $category->id = "$category->id,$category->contextid";
+            $newcatarray["$id,$category->contextid"] = $category;
+        }
+        return $newcatarray;
+    }
+
+    /**
+     * Finds top categories in the given categories hierarchy and replace their name with a proper localised string.
+     *
+     * @param array $categories An array of question categories.
+     * @param boolean $escape Whether the returned name of the thing is to be HTML escaped or not.
+     * @return array The same question category list given to the function, with the top category names being translated.
+     * @deprecated since Moodle 4.0 MDL-71585
+     * @see qbank_managecategories\helper
+     */
+    function question_fix_top_names($categories, $escape = true) {
```

Comment:
remove the deprecated code and point to the new function

Reply:
done, changed to public in managecategories.